### DISCUSSION TITLE
If empty [context] RAG should crash

### DISCUSSION THREAD

**Discussion by thiswillbeyourgithub** [open-webui/open-webui#5050] [General]
# Bug Report

## Installation Method

Docker compose

## Environment

- **Open WebUI Version:** 0.3.16

- **Operating System:** ubuntu 22
- **Browser (if applicable):** Brave

**Confirmation:**

- [ x ] I have read and followed all the instructions provided in the README.md.
- [ x ] I am on the latest version of both Open WebUI and Ollama.
- [ x ] I have included the browser console logs.
- [ x ] I have included the Docker container logs.
- [ x ] I have provided the exact steps to reproduce the bug in the "Steps to Reproduce" section below.

## Expected Behavior:

The expected behavior is that if the context is empty an error should happen.

## Actual Behavior:

No error happens and thr LLM just hallucinates answer.

## Description

**Bug Summary:**

It seems my RAG setup is broken. No matter what type of document I send, which browser I use etc looking at the traffic via langfuse shows me that the [context] is replaced by an empty string. The purpose of this issue is to add an exception when the context.strip() is empty.

## Reproduction Details

**Steps to Reproduce:**
Irrelevant.

**Comment by thiswillbeyourgithub** [comment]
Does the fact that the issue was moved to a discussion means the suggestion was refused? If so, can I kindly ask for an explanation? Because currently (and for at least like 4 weeks now) RAG and web search are silently broken because the template contains `<context></context>`.

Here are some of the things that I tried in the last weeks:
- switching embedder between ollama / sbert / openai
- turning on and off hybrid search
- Using or not using tikka
- Resetting vector storage / upload directory
- Changing the RAG template / resetting it to its default value
- Disabling all my functions and pipes and tools
- Trying with model hosted on ollama, or my litellm
- Any combination of the above

Seeing other  users potentially having similar issues (#5002 #5291 and maybe others), I can't help but think that (if I understood correctly) refusing to error out when the context is empty, artificially covers the problem : maybe many other users have a broken RAG but don't know it because LLMs are good at fooling us.

Edit: All the folks at #3363

**Reply by tjbck** [reply]
It means I cannot reproduce with any of my setup, let us know if you manage to find a fix!

**Reply by tjbck** [reply]
RAG:
<img width="1254" alt="image" src="https://github.com/user-attachments/assets/97f67717-0f5d-4c5b-8348-ba30f2bfc758">

Web Search:
<img width="1253" alt="image" src="https://github.com/user-attachments/assets/72234d4e-381f-4aed-91f0-35621bc64d6d">

**Reply by thiswillbeyourgithub** [reply]
Thanks a lot for the answer! (Open-WebUI is awesome and you can be proud of yourself if I may!)

I have not found a fix in weeks and I attribute this to the fact that it's very tedious to check my langfuse prompts / filter the logs where I think it should just error out.

Let me know if you would accept a PR that would add this [there](https://github.com/open-webui/open-webui/blob/e510c8b11f0f898796635cf09aa93a3d72c4c2ee/backend/open_webui/apps/rag/utils.py#L185):


``` diff
 def rag_template(template: str, context: str, query: str):
+    if not context.strip():
+        raise Exception("Failed to get context for RAG (received empty context)")
+    if template.count("[context]") != 1:
+        raise Exception("RAG template contains an unexpected number of '[context]' occurence")
     template = template.replace("[context]", context)
+    if "[query]" not in template:
+        raise Exception("RAG template is missing a [query] string")
     template = template.replace("[query]", query)
     return template
```

The expected result is that it would crash on my setup but not silently. It would also warn affected users where the crash is already happening but silently. This can only get us faster to the bottom of this.


PS: btw the current code does NOT handle situation where the `context` variable contains `[query]`. This way several [query] can be replaced. I think it should be replaced by a seed first. I can do that too in the PR if you want.



![image](https://github.com/user-attachments/assets/861f7715-7329-47ba-9eac-03f7ff67ebcd)

**Reply by thiswillbeyourgithub** [reply]
Hi @tjbck 

I fixed my issue. I'll submit a PR shortly.

**Reply by thiswillbeyourgithub** [reply]
Done at #5378
